# Data Exploration

## 1. Import libraries

In [1]:
import pandas as pd
import numpy as np

---

## 2. Load Raw Data

Trong giai đoạn đầu của quá trình khám phá dữ liệu, chúng em tiến hành tải dữ liệu thô (raw data) từ hai tệp CSV được cung cấp trong thư mục `data/raw`. 

Bộ dữ liệu thứ nhất, `worldometer_coronavirus_daily_data.csv`, chứa thông tin theo ngày về tình hình dịch COVID-19 tại từng quốc gia. 

Bộ dữ liệu thứ hai, `worldometer_coronavirus_summary_data.csv`, cung cấp số liệu tổng hợp cho từng quốc gia tại thời điểm thống kê, phản ánh tình hình dịch bệnh ở mức độ tổng quát hơn. 

### 2.1 Load daily COVID-19 dataset

In [2]:
daily_path = "../data/raw/worldometer_coronavirus_daily_data.csv"
df_daily = pd.read_csv(daily_path)

### 2.2 Load summary COVID-19 dataset

In [3]:
summary_path = "../data/raw/worldometer_coronavirus_summary_data.csv"
df_summary = pd.read_csv(summary_path)

## 3. Dataset Dimensions

### 3.1 Daily dataset

Dữ liệu theo ngày được đọc từ tệp `worldometer_coronavirus_daily_data.csv` và được lưu vào biến df_daily. Để có cái nhìn ban đầu về cấu trúc và đặc điểm của dataset, chúng em hiển thị 5 dòng đầu tiên và 5 dòng cuối cùng.

In [4]:
# Xem 5 dòng đầu
df_daily.head()

date      country  cumulative_total_cases  daily_new_cases  \
0  2020-2-15  Afghanistan                     0.0              NaN   
1  2020-2-16  Afghanistan                     0.0              NaN   
2  2020-2-17  Afghanistan                     0.0              NaN   
3  2020-2-18  Afghanistan                     0.0              NaN   
4  2020-2-19  Afghanistan                     0.0              NaN   

   active_cases  cumulative_total_deaths  daily_new_deaths  
0           0.0                      0.0               NaN  
1           0.0                      0.0               NaN  
2           0.0                      0.0               NaN  
3           0.0                      0.0               NaN  
4           0.0                      0.0               NaN

In [5]:
# Xem 5 dòng cuối
df_daily.tail()

date   country  cumulative_total_cases  daily_new_cases  \
184782  2022-5-10  Zimbabwe                248642.0            106.0   
184783  2022-5-11  Zimbabwe                248778.0            136.0   
184784  2022-5-12  Zimbabwe                248943.0            165.0   
184785  2022-5-13  Zimbabwe                249131.0            188.0   
184786  2022-5-14  Zimbabwe                249206.0             75.0   

        active_cases  cumulative_total_deaths  daily_new_deaths  
184782         963.0                   5481.0               2.0  
184783        1039.0                   5481.0               0.0  
184784        1158.0                   5481.0               0.0  
184785        1283.0                   5482.0               1.0  
184786        1307.0                   5482.0               0.0

Tính số dòng, số cột của dữ liệu

In [6]:
num_rows_df1, num_cols_df1 = df_daily.shape
print(f"Number of rows: {num_rows_df1}")
print(f"Number of columns: {num_cols_df1}")

Number of rows: 184787
Number of columns: 7


Dữ liệu theo ngày bao gồm 184787 dòng và 7 biến, phản ánh diễn biến dịch COVID-19 ở từng quốc gia theo từng ngày. Kích thước lớn của bộ dữ liệu cho thấy mức độ chi tiết cao theo chiều thời gian, có ý nghĩa đối với các phân tích xu hướng, tốc độ lan truyền, hoặc mô hình hóa dự báo.

### 3.2 Summary dataset

Dữ liệu tổng hợp được đọc từ tệp `worldometer_coronavirus_summary_data.csv` và được lưu vào biến df_summary. Tương tự, để có cái nhìn ban đầu về cấu trúc và đặc điểm của dataset, chúng em hiển thị 5 dòng đầu tiên và 5 dòng cuối cùng.

In [7]:
# Xem 5 dòng đầu
df_summary.head()

country continent  total_confirmed  total_deaths  total_recovered  \
0  Afghanistan      Asia           179267        7690.0         162202.0   
1      Albania    Europe           275574        3497.0         271826.0   
2      Algeria    Africa           265816        6875.0         178371.0   
3      Andorra    Europe            42156         153.0          41021.0   
4       Angola    Africa            99194        1900.0          97149.0   

   active_cases  serious_or_critical  total_cases_per_1m_population  \
0        9375.0               1124.0                           4420   
1         251.0                  2.0                          95954   
2       80570.0                  6.0                           5865   
3         982.0                 14.0                         543983   
4         145.0                  NaN                           2853   

   total_deaths_per_1m_population  total_tests  total_tests_per_1m_population  \
0                           190.0     951337.0                        23455.0   
1                          1218.0    1817530.0                       632857.0   
2                           152.0     230861.0                         5093.0   
3                          1974.0     249838.0                      3223924.0   
4                            55.0    1499795.0                        43136.0   

   population  
0    40560636  
1     2871945  
2    45325517  
3       77495  
4    34769277

In [8]:
# Xem 5 dòng cuối
df_summary.tail()

country          continent  total_confirmed  \
221  Wallis And Futuna Islands  Australia/Oceania              454   
222             Western Sahara             Africa               10   
223                      Yemen               Asia            11819   
224                     Zambia             Africa           320591   
225                   Zimbabwe             Africa           249206   

     total_deaths  total_recovered  active_cases  serious_or_critical  \
221           7.0            438.0           9.0                  NaN   
222           1.0              9.0           0.0                  NaN   
223        2149.0           9009.0         661.0                 23.0   
224        3983.0         315997.0         611.0                  NaN   
225        5482.0         242417.0        1307.0                 12.0   

     total_cases_per_1m_population  total_deaths_per_1m_population  \
221                          41755                           644.0   
222                             16                             2.0   
223                            381                            69.0   
224                          16575                           206.0   
225                          16324                           359.0   

     total_tests  total_tests_per_1m_population  population  
221      20508.0                      1886140.0       10873  
222          NaN                            NaN      624681  
223     265253.0                         8543.0    31049015  
224    3452554.0                       178497.0    19342381  
225    2287793.0                       149863.0    15265849

Tính số dòng, số cột của dữ liệu

In [9]:
num_rows_df2, num_cols_df2 = df_summary.shape
print(f"Number of rows: {num_rows_df2}")
print(f"Number of columns: {num_cols_df2}")

Number of rows: 226
Number of columns: 12


Bộ dữ liệu tổng hợp được tải từ tệp `worldometer_coronavirus_summary_data.csv`, bao gồm 226 dòng và 12 biến, đại diện cho số liệu thống kê của từng quốc gia tại một thời điểm tổng hợp. Quy mô nhỏ gọn của dataset nhưng chứa nhiều biến quan trọng giúp cung cấp cái nhìn toàn cảnh và hỗ trợ hiệu quả cho các phân tích so sánh giữa các quốc gia.

## 4. The meaning of each row

Trong hai bộ dữ liệu được sử dụng, mỗi dòng đều đại diện cho một đơn vị quan sát cụ thể nhưng có tính chất khác nhau tùy theo cấu trúc của từng file. 

Đối với **summary dataset**, mỗi dòng tương ứng với một quốc gia tại một thời điểm thống kê tổng hợp, bao gồm các chỉ số như tổng số ca nhiễm, tử vong, hồi phục, số ca đang điều trị, số ca nguy kịch, cùng nhiều chỉ số trên một triệu dân và quy mô dân số. Đây là dữ liệu "tĩnh", nghĩa là mỗi quốc gia xuất hiện đúng một lần và không mang tính chuỗi thời gian.

Ngược lại, trong **daily dataset**, mỗi dòng đại diện cho một cặp *quốc gia – ngày*, phản ánh số liệu COVID-19 theo thời gian, bao gồm tổng số ca tích lũy, ca mới trong ngày, số ca đang điều trị, tổng số tử vong và tử vong mới. Dữ liệu này có tính "động" và ghi nhận sự thay đổi hàng ngày của dịch bệnh. 

Việc so sánh hai tập dữ liệu cho thấy rằng ý nghĩa của mỗi dòng là nhất quán trong từng dataset, và không xuất hiện tình trạng dòng dữ liệu mang ý nghĩa khác nhau trong cùng một file. Sự khác biệt chỉ nằm ở cấp độ ghi nhận (daily và summary), điều này hoàn toàn phù hợp với mục tiêu và cấu trúc của từng loại dữ liệu.

## 5. Checking duplicated data

Trong quá trình kiểm tra tính nhất quán của dữ liệu, ta sử dụng phương thức `duplicated()` kết hợp với `any()` của Pandas để phát hiện các dòng trùng lặp. Biến lưu kết quả kiểm tra (`duplicated_daily` và `duplicated_summary`) có giá trị **True** nếu tồn tại các dòng trùng lặp, và **False** nếu dữ liệu hoàn toàn duy nhất.

In [14]:
duplicated_daily = df_daily.duplicated(subset = ["country", "date"]).any()
print(duplicated_daily)

False


Với **daily dataset**, mỗi dòng đại diện cho một cặp *(country, date)*. Việc kiểm tra trùng lặp theo hai cột này trả về giá trị `False`, cho thấy không có dòng nào bị lặp lại trong dữ liệu theo ngày. Điều này xác nhận rằng dữ liệu ghi nhận theo ngày cho mỗi quốc gia là duy nhất và không bị trùng nhập.

In [15]:
duplicated_summary = df_summary.duplicated(subset = ["country"]).any()
print(duplicated_summary)

False


Tương tự, với **summary dataset**, mỗi dòng đại diện cho một quốc gia duy nhất. Kiểm tra trùng lặp theo cột `country` cũng trả về giá trị `False`, đồng nghĩa với việc không có quốc gia nào xuất hiện nhiều lần trong dataset tổng hợp, và tổng số dòng trùng lặp bằng 0.

## 6. The meaning of each column

### 6.1 Daily dataset

In [18]:
df_daily.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 184787 entries, 0 to 184786
Data columns (total 7 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   date                     184787 non-null  object 
 1   country                  184787 non-null  object 
 2   cumulative_total_cases   184787 non-null  float64
 3   daily_new_cases          174329 non-null  float64
 4   active_cases             166747 non-null  float64
 5   cumulative_total_deaths  178227 non-null  float64
 6   daily_new_deaths         157850 non-null  float64
dtypes: float64(5), object(2)
memory usage: 9.9+ MB


* `date` (*object*): Ngày ghi nhận số liệu.
* `country` (*object*): Tên quốc gia; mỗi dòng tương ứng với một quốc gia tại một ngày cụ thể.
* `cumulative_total_cases` (*float64*): Tổng số ca nhiễm tích lũy đến ngày đó.
* `daily_new_cases` (*float64*): Số ca nhiễm mới trong ngày.
* `active_cases` (*float64*): Số ca đang điều trị.
* `cumulative_total_deaths` (*float64*): Tổng số ca tử vong tích lũy.
* `daily_new_deaths` (*float64*): Số ca tử vong mới trong ngày.

Dữ liệu `daily` có tính chất **chuỗi thời gian**, với mỗi dòng là một quan sát theo ngày của một quốc gia.

### 6.2 Summary dataset


In [21]:
df_summary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226 entries, 0 to 225
Data columns (total 12 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   country                         226 non-null    object 
 1   continent                       226 non-null    object 
 2   total_confirmed                 226 non-null    int64  
 3   total_deaths                    218 non-null    float64
 4   total_recovered                 204 non-null    float64
 5   active_cases                    204 non-null    float64
 6   serious_or_critical             145 non-null    float64
 7   total_cases_per_1m_population   226 non-null    int64  
 8   total_deaths_per_1m_population  218 non-null    float64
 9   total_tests                     212 non-null    float64
 10  total_tests_per_1m_population   212 non-null    float64
 11  population                      226 non-null    int64  
dtypes: float64(7), int64(3), object(2)
m

* `country` (*object*): Tên quốc gia; mỗi dòng là một quốc gia duy nhất.
* `continent` (*object*): Châu lục mà quốc gia thuộc về.
* `total_confirmed` (*int64*): Tổng số ca mắc COVID-19 tích lũy.
* `total_deaths` (*float64*): Tổng số ca tử vong.
* `total_recovered` (*float64*): Tổng số ca hồi phục.
* `active_cases` (*float64*): Số ca đang điều trị hiện tại.
* `serious_or_critical` (*float64*): Số ca nghiêm trọng hoặc nguy kịch.
* `total_cases_per_1m_population` (*int64*): Số ca tích lũy trên 1 triệu dân.
* `total_deaths_per_1m_population` (*float64*): Số ca tử vong trên 1 triệu dân.
* `total_tests` (*float64*): Tổng số xét nghiệm COVID-19 đã thực hiện.
* `total_tests_per_1m_population` (*float64*): Số xét nghiệm trên 1 triệu dân.
* `population` (*int64*): Dân số của quốc gia.

Dataset `summary` có tính chất **tổng hợp**, cung cấp cái nhìn toàn cảnh về mức độ ảnh hưởng của COVID-19 tại từng quốc gia.

## 7. The current data type of each column? Are there any columns with inappropriate data types that can be processed further?

### 7.1 Daily dataset

Cột `date` hiện đang ở dạng object, nghĩa là được lưu trữ như chuỗi văn bản nên cần chuyển từ kiểu dữ liệu dạng chuỗi sang datetime.

Cột `country`: kiểu dữ liệu dạng object, nghĩa là được lưu trữ như chuỗi văn bản, đây là kiểu dữ liệu phù hợp.

Tất cả các cột số liệu (`cumulative_total_cases`, `daily_new_cases`, `active_cases`, `cumulative_total_deaths`, `daily_new_deaths`): đều mang kiểu float64. Đây là kiểu phù hợp, không cần chuyển đổi thêm.


### 7.2 Summary dataset


Các cột định danh như `country` hoặc `iso_code` được lưu dưới dạng string (object), đây là kiểu dữ liệu đúng và phù hợp vì chúng mang thông tin phân loại. 

Các biến định lượng tổng hợp như `total_cases`, `total_deaths`, `population`, hoặc các chỉ số kinh tế - xã hội khác được kỳ vọng có kiểu numeric (integer hoặc float). Phần lớn các cột đều đang ở kiểu dữ liệu chính xác theo bản chất của biến.
